cmds/core/grep/grep.go

```diff
@@ -204,68 +199,85 @@ func (c *cmd) run() error {
-		// generate a chan of file names, bounded by the size of the chan. This in turn
-		// throttles the opens.
-		treeNames := make(chan string, 128)
-		go func() {
-			defer close(treeNames)
-			for _, v := range c.args[1:] {
-				filepath.Walk(v, func(name string, fi os.FileInfo, err error) error {
-					if err != nil {
-						fmt.Fprintf(c.stderr, "grep: %v: %v\n", name, err)
-						return nil
-					}
-					if fi.IsDir() && !c.recursive {
-						fmt.Fprintf(c.stderr, "grep: %v: Is a directory\n", name)
-						return filepath.SkipDir
-					}
-					treeNames <- name
-					return nil
-				})
-			}
-		}()
 
-		files := make(chan *grepCommand)
-		// convert the file names to a stream of os.File
-		go func() {
-			for i := range treeNames {
-				fp, err := os.Open(i)
+	// start producing greps in a goroutine
+	go func() {
+		// very special case, just stdin
+		if len(c.args) < 2 {
+			c.run_stdin(re)
+		} else {
+			c.run_files(re)
+		}
+		// the allGrep channel is closed when work is done
+		close(c.allGrep)
+	}()
+
+	return c.run_print()
+}
+
+func (c *cmd) run_stdin(re *regexp.Regexp) {
+	res := make(chan *grepResult, 1)
+	go c.grep(&grepCommand{c.stdin, "<stdin>"}, re, res)
+	c.allGrep <- &oneGrep{res}
+
+}
+
+func (c *cmd) run_files(re *regexp.Regexp) {
+	c.showName = (len(c.args[1:]) > 1 || c.recursive || c.noShowMatch) && !c.headers
+	// generate a chan of file names, bounded by the size of the chan. This in turn
+	// throttles the opens.
+	treeNames := make(chan string, 128)
+	go func() {
+		defer close(treeNames)
+		for _, v := range c.args[1:] {
+			filepath.Walk(v, func(name string, fi os.FileInfo, err error) error {
 				if err != nil {
-					fmt.Fprintf(c.stderr, "can't open %s: %v\n", i, err)
-					continue
+					fmt.Fprintf(c.stderr, "grep: %v: %v\n", name, err)
+					return nil
 				}
-				files <- &grepCommand{fp, i}
-			}
-			close(files)
-		}()
-		// now kick off the greps
-		// bug: file name order is not preserved here. Darn.
+				if fi.IsDir() && !c.recursive {
+					fmt.Fprintf(c.stderr, "grep: %v: Is a directory\n", name)
+					return filepath.SkipDir
+				}
+				treeNames <- name
+				return nil
+			})
+		}
+	}()
 
-		for f := range files {
-			c.nGrep++
-			go c.grep(f, re)
+	files := make(chan *grepCommand)
+	// convert the file names to a stream of os.File
+	go func() {
+		for i := range treeNames {
+			fp, err := os.Open(i)
+			if err != nil {
+				fmt.Fprintf(c.stderr, "can't open %s: %v\n", i, err)
+				continue
+			}
+			files <- &grepCommand{fp, i}
 		}
+		close(files)
+	}()
+	// now kick off the greps
+	for f := range files {
+		res := make(chan *grepResult, 1)
+		c.allGrep <- &oneGrep{res}
+		go c.grep(f, re, res)
 	}
+}
 
-	if c.nGrep > 0 {
-		for og := range c.allGrep {
-			for r := range og.c {
-				// exit on first match.
-				if c.quiet {
-					return nil
-				}
-				c.printMatch(r)
-			}
-			c.nGrep--
-			if c.nGrep == 0 {
-				break
+func (c *cmd) run_print() error {
```

Comment:
I don't think snake case is the right way of writing idiomatic Go code.  Please take a look at the other functions as well.

By the way, I love snake case.

Reply:
ahhh yeah in my org we use snake case for single use unexported functions and lowerCamelCase for multiple use. 

i can change everything (but we should probably reach a decision on the concurrency thing)